lib/adm.js

```diff
@@ -4127,34 +4127,188 @@ maAdm.prototype.generatePlan = function (opts, callback)
+	 */
+
+	var imgapi_opts = {};
+	if (opts.channel !== null) {
+		imgapi_opts.channel = opts.channel;
+	}
+
+	vasync.forEachParallel({
+		inputs: Object.keys(opts.services_by_image),
+		func: function fetchOneImage(imageUuid, nextImage) {
+		opts.log.debug(
+				'getting remote image %s for service %s on ' +
+				'channel %s',
+				imageUuid, opts.services_by_image[imageUuid],
+				opts.channel);
+		opts.remote_imgapi.getImage(imageUuid, imgapi_opts,
+			function getImage(err, image) {
+			if (err) {
+				var missing = sprintf('%s (%s)', imageUuid,
+				    opts.services_by_image[imageUuid]);
+				opts.log.debug(
+				    'missing remote image ' + missing);
+				nextImage();
+			} else {
+				opts.remote_images[imageUuid] = image;
+				nextImage();
+			}
+		});
+		}
+	}, function finishFetchRemoteImages(err) {
+		if (err) {
+			opts.log.debug('error getting remote images %d', err);
+			cb(err);
+		} else {
+			cb();
+		}
+	});
+}
+
+/*
+ * Retrieves Image objects, invoking fetchLocalImages and fetchRemoteImages.
+ * Missing images are not considered errors at this point.
+ */
+function fetchImages(opts, cb) {
+	assertplus.object(opts, 'opts');
+	assertplus.func(cb, 'cb');
+	assertplus.object(opts.services_by_image, 'opts.services_by_image');
+	assertplus.object(opts.remote_imgapi, 'opts.remote_imgapi');
+	assertplus.object(opts.local_imgapi, 'opts.local_imgapi');
+	assertplus.bool(opts.skip_verify_remote, 'opts.skip_verify_remote');
+
+	vasync.pipeline({
+		arg: opts,
+		funcs: [
+			function stepFetchLocalImages(arg, stepcb) {
+				fetchLocalImages(arg, stepcb);
+			},
+			// even if we've got local images, we'll still need the
+			// remote ones in order to verify the channel
+			function stepFetchRemoteImages(arg, stepcb) {
+				fetchRemoteImages(arg, stepcb);
+			}
+		]
+	}, function cbImagesLists(err) {
+		if (err) {
+			cb(err);
+			return;
+		} else {
+			cb(null, opts.local_images, opts.remote_images);
+		}
+	});
+}
+
 /*
  * Determine whether the plan that was generated adheres to any restrictions
  * that are in place.
- * For now the only check implemented is to ensure that all images for
- * deployment are available on the specific imgapi channel set by
- * 'sdcadm channel set' on the 'remote_imgapi' set in etc/config.json.
  *
- * Options:
- * - skip_verify_channel: if true, we avoid verifying that each image to be
+ * We check the following restrictions:
+ *
+ * 1. All images for deployment are available on the specific imgapi channel
+ *    set by 'sdcadm channel set' on the 'remote_imgapi' set in etc/config.json.
+ *
+ * 2. All images for deployment have image names that match the service they're
+ *    being deployed for. It would be bad if were to able to reprovision a
+ *    moray zone with anything other than a moray image.
+ *
+ * 3. All images for deployment exist in the local imgapi instance in this
+ *    datacenter
+ *
+ *  Options:
+ * * - skip_verify_channel: if true, we avoid verifying that each image to be
```

Comment:
nit: Why an asterisk *and* a dash for double-bullets there?

Reply:
Och, sorry yes, of course.